cts/scheduler/summary/cancel-behind-moving-remote.summary

```diff
@@ -210,10 +185,10 @@ GuestOnline: [ galera-bundle-0:galera-bundle-podman-0 galera-bundle-1:galera-bun
    haproxy-bundle-podman-1	(ocf:heartbeat:podman):	 Started controller-0
    haproxy-bundle-podman-2	(ocf:heartbeat:podman):	 Started controller-1
  Container bundle set: ovn-dbs-bundle [cluster.common.tag/rhosp16-openstack-ovn-northd:pcmklatest]
-   ovn-dbs-bundle-0	(ocf:ovn:ovndb-servers):	 Slave controller-2
-   ovn-dbs-bundle-1	(ocf:ovn:ovndb-servers):	 Master controller-0
+   ovn-dbs-bundle-0	(ocf:ovn:ovndb-servers):	 Slave controller-0
+   ovn-dbs-bundle-1	(ocf:ovn:ovndb-servers):	 Master controller-2
    ovn-dbs-bundle-2	(ocf:ovn:ovndb-servers):	 Slave controller-1
- ip-172.17.1.87	(ocf:heartbeat:IPaddr2):	 Started controller-0
+ ip-172.17.1.87	(ocf:heartbeat:IPaddr2):	 Stopped
```

Comment:
To save you some potential confusion during review: The ip-172.17.1.87 resource is now stopped because it's banned from node 2.

Reply:
Hmm, I'm not sure that's acceptable. Selecting a node for the promotion of ovn-dbs-bundle should take into account colocation dependencies' preferences. Previously that was implicit by the ordering of the clone instances since the first N nodes are chosen for promotion. I'm not sure what the resolution is, maybe we need to re-sort the nodes before setting instance roles.

Reply:
Good catch. I thought the behavior would have been the same on the existing scheduler if the ban constraint were for `controller-0` instead, but I'm testing now and that seems not to be the case.

The fact that `ovn-dbs-bundle-0` is going to `controller-0` and `ovn-dbs-bundle-1` is going to `controller-2` now is good and we want to keep that. It doesn't make sense to move instance 1. As you said, maybe there's something we can do in `promotion_order()`.

Sigh.

Reply:
In `pcmk__set_instance_roles()`, there's a block that looks like it's meant to apply constraints, but it doesn't find any.
```
        for (gIter2 = child_rsc->rsc_cons; gIter2 != NULL; gIter2 = gIter2->next) {
            pcmk__colocation_t *cons = (pcmk__colocation_t *) gIter2->data;

            child_rsc->cmds->rsc_colocation_lh(child_rsc, cons->rsc_rh, cons,
                                               data_set);
        }
```
```
(gdb) p rsc->id
$9 = 0x13e0bd0 "ovn-dbs-bundle-master"
(gdb) p child_rsc->id
$10 = 0x13f59b0 "ovndb_servers:1"
(gdb) p rsc->rsc_cons
$11 = 0x0
(gdb) p rsc->rsc_cons_lhs
$12 = 0x0
(gdb) p child_rsc->rsc_cons
$13 = 0x0
(gdb) p child_rsc->rsc_cons_lhs
$14 = 0x0
```

I might be misunderstanding how constraints are stored. Posting this here for now in case I don't get a chance to track it down.

Reply:
Okay, scratch the constraints. the direct cause of this unwanted behavior isn't that the colocation constraints aren't being honored. The direct cause is that `ovndb_servers:0` is in Stopped state when we begin. This causes a few issues.

First, `lookup_promotion_score()` is unable to get the promotion score for `ovndb_servers:0`.
```
(pe_node_attribute_calculated)  trace: ovndb_servers:0: Not looking for master-ovndb_servers:0 on the container host: ovn-dbs-bundle-podman-0 is inactive
```

That's because `pe_node_attribute_calculated()` checks the first node in the container's `running_on` list. If we modify it as follows:
```
    //if (node->details->remote_rsc->container->running_on != NULL) {
        //pe_node_t *host = node->details->remote_rsc->container->running_on->data;
    if (node->details->remote_rsc->container->allocated_to != NULL) {
        pe_node_t *host = node->details->remote_rsc->container->allocated_to;
```

Then we're able to get the promotion score for `ovndb_servers:0`. (Whether the above will have any unwanted side effects, I don't know yet. We could also check `allocated_to` and then `running_on`, or vice versa.)
```
(pe_node_attribute_calculated)  trace: ovndb_servers:0: Looking for master-ovndb_servers:0 on the allocated container host controller-0
(promotion_score)       trace: promotion score for ovndb_servers:0 on ovn-dbs-bundle-0 = <null>
(pe_node_attribute_calculated)  trace: ovndb_servers:0: Looking for master-ovndb_servers on the allocated container host controller-0
(promotion_score)       trace: stripped promotion score for ovndb_servers on ovn-dbs-bundle-0 = 5
```

So now we have the promotion score. Our next issue is that `sort_promotable_instance()` sorts based on the current role.
```
    role1 = resource1->fns->state(resource1, TRUE);
    role2 = resource2->fns->state(resource2, TRUE);

    if (role1 > role2) {
        crm_trace("%s %c %s (role)", resource1->id, '<', resource2->id);
        return -1;

    } else if (role1 < role2) {
        crm_trace("%s %c %s (role)", resource1->id, '>', resource2->id);
        return 1;
    }
```

If we modify that to sort by next role (again, haven't checked whether that has any unwanted side effects), then we encounter a third issue: `ovndb_servers:0` is inactive, so it loses in the sort order to the active instances.

---
On a positive note: If the `pe_node_attribute_calculated()` fix is in place, then on the second transition, the correct node gets promoted and the IP starts. It's able to do this because at that point, `ovndb_servers:0` is already in started state.
```
# crm_simulate -S -x cts/scheduler/xml/cancel-behind-moving-remote.xml -O /tmp/next.xml >/dev/null
# crm_simulate -R -x /tmp/next.xml
...
Transition Summary:
 * Start      rabbitmq-bundle-1     (                     controller-0 )   due to unrunnable rabbitmq-bundle-podman-1 start (blocked)
 * Start      rabbitmq:1            (                rabbitmq-bundle-1 )   due to unrunnable rabbitmq-bundle-podman-1 start (blocked)
 * Promote    ovndb_servers:0       ( Slave -> Master ovn-dbs-bundle-0 )  
 * Demote     ovndb_servers:1       ( Master -> Slave ovn-dbs-bundle-1 )  
 * Start      ip-172.17.1.87        (                     controller-0 )
```

So colocation constraints are being accounted for.

---
To summarize, I feel like it would be fine to check the `allocated_to` node in `pe_node_attribute_calculated()` as a simple fix for looking up the promotion score. But the remaining issue is "how do we sort the promotable instances so that a Stopped instance has a fair chance to be promoted **on the first transition**, without making the sort criteria problematic for other situations?"

P.S. The reason this worked before the fix is that the stopped instance (`ovndb_servers:0`) was allocated to node that didn't need to be promoted. Now, the stopped instance is allocated to `controller-0`, which is the node that needs to be promoted so that the IP can start. The instance on `controller-2` gets promoted instead, and the IP can't start there.

Reply:
> My understanding is that `rsc_cons_lhs` is the list of constraints where some other resource depends on this resource... and **not** the list of constraints for which this resource is on the left-hand side. I bet I'm not the only one who gets a headache from the lh/rh names.

I know right! We can't change struct member names until we're ready to break backward API compatibility, but I have it on my to-do list for when that happens: s/rsc_cons/this_with/ and s/rsc_cons_lhs/with_this/



> First, `lookup_promotion_score()` is unable to get the promotion score for `ovndb_servers:0`.
> 
> ```
> (pe_node_attribute_calculated)  trace: ovndb_servers:0: Not looking for master-ovndb_servers:0 on the container host: ovn-dbs-bundle-podman-0 is inactive
> ```
> 
> That's because `pe_node_attribute_calculated()` checks the first node in the container's `running_on` list. If we modify it as follows:
> 
> ```
>     //if (node->details->remote_rsc->container->running_on != NULL) {
>         //pe_node_t *host = node->details->remote_rsc->container->running_on->data;
>     if (node->details->remote_rsc->container->allocated_to != NULL) {
>         pe_node_t *host = node->details->remote_rsc->container->allocated_to;
> ```
> 
> Then we're able to get the promotion score for `ovndb_servers:0`. (Whether the above will have any unwanted side effects, I don't know yet. We could also check `allocated_to` and then `running_on`, or vice versa.)

pe_node_attribute_calculated() is different from pe_node_attribute_raw() only for resources in a bundle where container-attribute-target has been set to "host" (indicating that the value for the bundle instance's host should be used rather than any value set for the bundle instance node itself).

Besides promotion_score(), calculated attributes are used for location rules where the rule score is taken from a node attribute, for example to say a containerized resource has to run on a host with a node attribute indicating it has access to necessary shared storage.

Location rules feed into the allocation process, so we don't have ->allocated_to at that point.

Offhand, it sounds reasonable to use ->allocated_to if it's non-NULL and ->running_on otherwise -- it should work in practice for these two purposes. Alternatively we could pass a bool to pe_node_attribute_calculated() indicating which one we want. Hopefully existing regression tests are sufficient to rule out side effects.

Reply:
> Offhand, it sounds reasonable to use ->allocated_to if it's non-NULL and ->running_on otherwise -- it should work in practice for these two purposes.

It's worth pointing out that pcmk__set_instance_roles() calls promotion_score() with the ->allocated_to node (via ->location(..., FALSE)), so it's reasonable to assume the host's ->allocated_to is the right choice for that context.

Reply:
> Our next issue is that `sort_promotable_instance()` sorts based on the current role.
> 
> ```
>     role1 = resource1->fns->state(resource1, TRUE);
>     role2 = resource2->fns->state(resource2, TRUE);
> 
>     if (role1 > role2) {
>         crm_trace("%s %c %s (role)", resource1->id, '<', resource2->id);
>         return -1;
> 
>     } else if (role1 < role2) {
>         crm_trace("%s %c %s (role)", resource1->id, '>', resource2->id);
>         return 1;
>     }
> ```
> 
> If we modify that to sort by next role (again, haven't checked whether that has any unwanted side effects),

The point of promotion_order() is to decide which instances to promote, so at this point none of the children will have a next_role of promoted. Using the current role allows current promoted instances to be favored to keep their role. Also, preferring already-started instances over to-be-started instances probably has some value in terms of getting a promoted instance more quickly.

Before comparing roles, sort_promotable_instance() compares rsc->sort_index (via sort_rsc_index()). It feels like we need either a new criteria between those two, or to adjust ->sort_index in promotion_order().

Reply:
I had to take a detour and get a better grasp on how bundles and colocation work, before I could add what's most likely going to be a simple fix. I've got a version that appears to work now. It needs some cleanup. Might have something by tomorrow.